File: Agents/python/LightingAgent/utilities/assistant_event_stream_utility.py

import json
from typing import Optional
from pydantic import BaseModel
from models.assistant_message_content import AssistantMessageContent
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Generator, Optional
from openai.types.chat.chat_completion_chunk import ChatCompletionChunk
from semantic_kernel.contents.streaming_chat_message_content import StreamingChatMessageContent
from semantic_kernel.contents import TextContent, AuthorRole


@dataclass
class StreamingChatCompletionsUpdate:
    Id: str
    FinishReason: Optional[str]

class AssistantEventStreamUtility:
    _current_message: Optional[AssistantMessageContent] = None

    def create_message_event(self, run_id: str, data: StreamingChatMessageContent) -> Generator[str, None, None]:
        streaming_chat_completions_update: ChatCompletionChunk = data.InnerContent

        if (self._current_message is not None and streaming_chat_completions_update.id != self._current_message.id):
            yield self.create_error_event("Previous message was not finished.")
            return

        if self._current_message is None:
            self._current_message = AssistantMessageContent(
                id=streaming_chat_completions_update.id,
                thread_id=self._current_thread_id,
                created_at=datetime.now(),
                role=AuthorRole.Assistant,
                assistant_id="LightingAgent",
                run_id=run_id,
                items=[]
            )
            yield self.create_event("thread.message.created", self._current_message)
            yield self.create_event("thread.message.in_progress", self._current_message)

        self._message_builder.append(data.content)

        yield self.create_event("thread.message.delta", data)

        if streaming_chat_completions_update.choices[0].finish_reason is not None:
            self._current_message.items = [TextContent(''.join(self._message_builder))]
            yield self.create_event("thread.message.completed", self._current_message)
            self._current_message = None

    def create_event(self, event_type: str, data: BaseModel) -> str:
        json_data = data.model_dump_json()
        return f"event: {event_type}\n" + f"data: {json_data}\n\n"

    def create_error_event(self, message: str) -> str:
        error_data = {"message": message}
        json_data = json.dumps(error_data)
        return f"event: error\n" + f"data: {json_data}\n\n"

    def create_done_event(self) -> str:
        return "event: done\n" + "data: [DONE]\n\n"

File: Agents/dotnet/Shared/Utilities/AssistantEventStreamUtility.cs

using System.Text;
using System.Text.Json;
using Azure.AI.OpenAI;
using Microsoft.AspNetCore.Http.Json;
using Microsoft.Extensions.Options;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.OpenAI;
using Shared.Config;
using Shared.Models;

namespace Shared.Utilities
{
    public class AssistantEventStreamUtility(
        IOptions<AgentConfig> agentConfiguration,
        IOptions<Microsoft.AspNetCore.Mvc.JsonOptions> jsonOptions
    )
    {
        private string? _currentThreadId = null;
        private AssistantMessageContent? _currentMessage = null;
        private AgentConfig _agentConfiguration = agentConfiguration.Value;
        private StringBuilder _messageBuilder = new();

        public IEnumerable<string> CreateMessageEvent(string runId, StreamingChatMessageContent data)
        {
            // Check if the type is not OpenAIStreamingChatMessageContent (if so, fail)
            if (data.GetType() != typeof(OpenAIStreamingChatMessageContent))
            {
                yield return CreateErrorEvent("Only OpenAI chat completion APIs are supported.");
                yield break;
            }

            // Get the inner content
            var streamingChatCompletionsUpdate = (StreamingChatCompletionsUpdate)data.InnerContent!;

            // Check if the message was not closed properly
            if (_currentMessage != null && streamingChatCompletionsUpdate.Id != _currentMessage.Id)
            {
                yield return CreateErrorEvent("Previous message was not finished.");
                yield break;
            }

            // Check if the message is new
            if (_currentMessage == null)
            {
                // Create new message object
                _currentMessage = new AssistantMessageContent()
                {
                    Id = streamingChatCompletionsUpdate.Id,
                    ThreadId = _currentThreadId,
                    CreatedAt = DateTime.Now,
                    Role = AuthorRole.Assistant,
                    AssistantId = _agentConfiguration.Name,
                    RunId = runId,
                    Items = []
                };

                // Create new message events
                yield return CreateEvent("thread.message.created", _currentMessage);
                yield return CreateEvent("thread.message.in_progress", _currentMessage);
            }

            // Add the message delta to the complete message
            _messageBuilder.Append(data.Content);

            // Create message delta event
            yield return CreateEvent("thread.message.delta", data);

            // Check if the message is done
            if (streamingChatCompletionsUpdate.FinishReason != null)
            {
                // Update the message content
                _currentMessage.Items = [new TextContent(_messageBuilder.ToString())];

                // Create done message events
                yield return CreateEvent("thread.message.completed", _currentMessage);
                _currentMessage = null;
            }
        }

        public string CreateEvent<T>(string eventType, T data)
        {
            string jsonData = JsonSerializer.Serialize(data, options: jsonOptions.Value.JsonSerializerOptions);
            return $"event: {eventType}\n" +
                $"data: {jsonData}\n\n";
        }

        public string CreateErrorEvent(string message)
        {
            var errorData = new { message = message };
            string jsonData = JsonSerializer.Serialize(errorData, options: jsonOptions.Value.JsonSerializerOptions);
            return $"event: error\n" +
                $"data: {jsonData}\n\n";
        }

        public string CreateDoneEvent()
        {
            return "event: done\n" +
                "data: [DONE]\n\n";
        }
    }

}